sdk/iot/core/tests/cmocka/az_iot_telemetry_tests.c

```diff
@@ -0,0 +1,301 @@
+static const char g_test_correct_topic_with_options_module_id_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_user_agent_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS
+                                "&" TEST_USER_AGENT;
+
+static az_iot_hub_client g_test_valid_client_no_options
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_EMPTY_OPTIONS } };
+static const az_iot_hub_client g_test_valid_client_with_options_both
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_BOTH } };
+static const az_iot_hub_client g_test_valid_client_with_options_module_id
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_MODULE_ID } };
+static const az_iot_hub_client g_test_valid_client_with_options_user_agent
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_USER_AGENT } };
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_client_fails(void** state)
+{
+  (void)state;
+
+  uint8_t mqtt_topic_buf[TEST_MQTT_SPAN_BUFFER_SIZE];
+
+  az_span mqtt_topic = AZ_SPAN_LITERAL_FROM_BUFFER(mqtt_topic_buf);
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(NULL, NULL, mqtt_topic, &mqtt_topic)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_mqtt_topic_fails(void** state)
+{
+  (void)state;
+  az_span null_mqtt_topic = TEST_NULL_PTR_SPAN;
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(
+          &g_test_valid_client_no_options, NULL, null_mqtt_topic, &null_mqtt_topic)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_out_mqtt_topic_fails(void** state)
+{
+  (void)state;
+  uint8_t mqtt_topic_buf[TEST_MQTT_SPAN_BUFFER_SIZE];
+
+  az_span mqtt_topic = AZ_SPAN_LITERAL_FROM_BUFFER(mqtt_topic_buf);
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(
+          &g_test_valid_client_no_options, NULL, mqtt_topic, NULL)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_no_options_no_params_succeed(void** state)
+{
+  (void)state;
+
+  uint8_t mqtt_topic_buf[sizeof(g_test_correct_topic_no_options_no_params)];
```

Comment:
>sizeof(g_test_correct_topic_no_options_no_params) [](start = 25, length = 49)

this should return 4, is that what you want?

Reply:
nope since it's an array it returns the full size

Reply:
sizeof should return size of type and in this case its pointer so it should return 4 in 32 bit compiler and 8 in 64 bit compiler. If you have used macro then I am sure it will return you right size. Check a simple I create, hope that make my point clear.

[https://onlinegdb.com/B14Y-PeB8](https://onlinegdb.com/B14Y-PeB8)

---
In reply to: [389183249](https://github.com/CIPop/azure-sdk-for-c/pull/1#discussion_r389183249) [](ancestors = 389183249)

Reply:
In this case it isn't though. I changed it to mimic what I have and get the correct behavior:
[Link](https://onlinegdb.com/HJ7OcDlHL)